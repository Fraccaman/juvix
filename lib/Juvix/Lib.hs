{-# LANGUAGE EmptyDataDecls         #-}
{-# LANGUAGE ExplicitForAll         #-}
{-# LANGUAGE FlexibleInstances      #-}
{-# LANGUAGE FunctionalDependencies #-}
{-# LANGUAGE MagicHash              #-}
{-# LANGUAGE MultiParamTypeClasses  #-}
{-# LANGUAGE UnicodeSyntax          #-}

{-  Prelude-equivalent for Juvix-compiled programs.
    This file is mostly a lie; these functions are replaced by the transpiler with Michelson builtins.
    The intent of this library is to be "batteries-included" - so additional common functionality should be added as necessary.
    Subject to API instability until Michelson is finalized.  -}

module Juvix.Lib (

  {- Datatypes -}

  List,
  Set,
  Map,
  Signature,
  Key,
  Contract,
  Account,
  Either(..),
  Maybe,
  Pair,
  Unit,
  Tez,
  String,
  Int,
  Nat,
  Bool(..),

  {- General -}

  IO,
  Applicative,
  Functor,
  Monad,
  Monoid,
  pure,
  (<*>),
  fmap,
  (>>=),
  return,
  mappend,
  mempty,
  (<>),

  {- Boolean Operations -}

  (&&),
  (||),
  not,
  xor,

  {- Comparision -}

  Ord,
  compare,
  max,
  min,
  (<),
  (<=),
  (>),
  (>=),

  {- Equality -}

  Eq,
  (==),
  (/=),

  {- Arithmetic -}

  Mul,
  Add,
  Div,
  Sub,
  Negate,
  (+),
  (-),
  (*),
  (/),
  negate,
  abs,

  {- Michelson Calls -}

  manager,
  defaultAccount,
  source,
  self,
  transferTokens,
  now,
  amount,
  balance,
  hash,
  stepsToQuota,
  fail,

  {- Misc -}

  either,
  maybe,
  fromMaybe,
  emptySet,
  setMember,
  setUpdate,
  setSize,
  emptyMap,
  mapGet,
  mapMember,
  mapUpdate,
  mapSize

) where

import           Data.Either (Either (..), either)
import           Data.Eq     (Eq, (/=), (==))
import           Data.Maybe  (Maybe (..), fromMaybe, maybe)
import           Data.Monoid (Monoid, mappend, mempty, (<>))
import           Data.Ord    (Ord, compare, max, min, (<), (<=), (>), (>=))
import           Data.String (IsString, fromString)
import           GHC.Base    (Applicative, Functor, Monad, fmap, pure, return,
                              (<*>), (>>=))
import qualified GHC.Base    (fail)
import           GHC.Num     (Num, fromInteger)
import           GHC.Tuple   hiding (Unit)
import           GHC.Types   (Bool (..), IO)

class Mul a b c | a b → c where
  (*) ∷ a → b → c

class Add a b c | a b → c where
  (+) ∷ a → b → c

class Sub a b c | a b → c where
  (-) ∷ a → b → c

class Div a b c | a b → c where
  (/) ∷ a → b → c

class Negate a where
  negate  ∷ a → a
  abs     ∷ a → a

type List a = [a]

{-  May make sense to encapsulate common functionality between maps/sets in typeclasses. -}

data Set a

emptySet ∷ ∀ a . Set a
emptySet = michelson "EmptySetUT"

setMember ∷ ∀ a . a → Set a → Bool
setMember = michelson "SetMemUT"

setUpdate ∷ ∀ a . a → Bool → Set a → Set a
setUpdate = michelson "SetUpdateUT"

setSize ∷ ∀ a . Set a → Int
setSize = michelson "SetSizeUT"

data Map k v

emptyMap ∷ ∀ k v . (Ord k) ⇒ Map k v
emptyMap = michelson "EmptyMapUT"

mapGet ∷ ∀ k v . (Ord k) ⇒ k → Map k v → Maybe v
mapGet = michelson "MapGetUT"

mapMember ∷ ∀ k v . (Ord k) ⇒ k → Map k v → Bool
mapMember = michelson "MapMemberUT"

mapUpdate ∷ ∀ k v . (Ord k) ⇒ k → Maybe v → Map k v → Bool
mapUpdate = michelson "MapUpdateUT"

mapSize ∷ ∀ k v . (Ord k) ⇒ Map k v → Int
mapSize = michelson "MapSizeUT"

data Signature

checkSignature ∷ Key → (Signature, String) → Bool
checkSignature = michelson "CheckSignatureUT"

data Key

data Contract a b

type Account = Contract () ()

manager ∷ ∀ a b . Contract a b → Key
manager = michelson "ManagerUT"

defaultAccount ∷ Key → Account
defaultAccount = michelson "DefaultAccountUT"

source ∷ ∀ a b . Contract a b
source = michelson "SourceUT"

self ∷ ∀ a b . Contract a b
self = michelson "SelfUT"

{-  c *must* be the type of the contract's storage - at the moment this is not enfored by the Haskell typesystem  -}

transferTokens ∷ ∀ a b c . a → Tez → Contract a b → c → IO (b, c)
transferTokens = michelson "TransferTokensUT"

type Pair a b = (a, b)

type Unit = ()

unit ∷ Unit
unit = ()

data Timestamp

now ∷ Timestamp
now = michelson "NowUT"

data Nat

data Tez

amount ∷ Tez
amount = michelson "AmountUT"

balance ∷ Tez
balance = michelson "BalanceUT"

(&&) ∷ Bool → Bool → Bool
(&&) = michelson "AndUT"

(||) ∷ Bool → Bool → Bool
(||) = michelson "OrUT"

xor ∷ Bool → Bool → Bool
xor = michelson "XorUT"

not ∷ Bool → Bool
not = michelson "NotUT"

data String

hash ∷ a → String
hash = michelson "HUT"

data Int

{-  This is *not* a side effect - however, it makes no sense without specified evaluation order, so it's in the IO monad.
    Not sure if this is useful with Juvix anyways; if the quota is static the transpiler can analyze it.  -}

stepsToQuota ∷ IO Int
stepsToQuota = michelson "StepsToQuotaUT"

{-  Typeclass Instances   -}

{-  Equality      -}

instance Eq Tez where
  x == y = (michelson "CmpEqUT") x y

instance Eq Int where
  x == y = (michelson "CmpEqUT") x y

instance Eq Nat where
  x == y = (michelson "CmpEqUT") x y

{-  Comparision   -}

instance Ord Tez where
  x <= y = (michelson "CmpLeUT") x y

instance Ord Int where
  x <= y = (michelson "CmpLeUT") x y

instance Ord Nat where
  x <= y = (michelson "CmpLeUT") x y

{-  Arithmetic    -}

instance Add Int Int Int where
  (+) = michelson "AddIntIntUT"

instance Sub Int Int Int where
  (-) = michelson "SubIntUT"

instance Mul Int Int Int where
  (*) = michelson "MulIntIntUT"

instance Div Int Int Int where
  (/) = michelson "EdivIntIntUT"

instance Negate Int where
  negate  = michelson "NegIntUT"
  abs     = michelson "AbsIntUT"

instance Negate Nat where
  negate  = michelson "NegNatUT"
  abs     = michelson "AbsNatUT"

instance Add Nat Nat Nat where
  (+) = michelson "AddNatNatUT"

instance Sub Nat Nat Nat where
  (-) = michelson "SubNatNatUT"

instance Mul Nat Nat Nat where
  (*) = michelson "MulNatNatUT"

instance Div Nat Nat Nat where
  (/) = michelson "EdivNatNatUT"

instance Add Tez Tez Tez where
  (+) = michelson "AddTezUT"

instance Sub Tez Tez Tez where
  (-) = michelson "SubTezUT"

instance Add Int Nat Int where
  (+) = michelson "AddIntNatUT"

instance Add Nat Int Int where
  (+) = michelson "AddNatIntUT"

instance Mul Int Nat Int where
  (*) = michelson "MulIntNatUT"

instance Mul Nat Int Int where
  (*) = michelson "MulNatIntUT"

{-  Strings   -}

instance Monoid String where
  mempty  = fromString ""
  mappend = michelson "ConcatUT"

{-  For string literal overloading.   -}

instance IsString String where
  fromString = rewrite "stringFromString"

{-  Never exposed; we need these for numeric literal overloading.   -}

instance Num Tez where
  fromInteger = rewrite "tezFromInteger"
  {-# NOINLINE fromInteger #-}

instance Num Int where
  fromInteger = rewrite "intFromInteger"
  {-# NOINLINE fromInteger #-}

instance Num Nat where
  fromInteger = rewrite "natFromInteger"
  {-# NOINLINE fromInteger #-}

fail ∷ IO a
fail = michelson "FailUT"

{-  Michelson library calls, rewritten by Juvix.  -}

michelson ∷ ∀ a b . a → b
michelson op = let x = x in x
{-# NOINLINE michelson #-}

{-  Auxiliary rewrites.   -}

rewrite ∷ ∀ a b . a → b
rewrite label = let x = x in x
{-# NOINLINE rewrite #-}
