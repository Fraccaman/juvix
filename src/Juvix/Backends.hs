module Juvix.Backends where

import           Juvix.Backends.Michelson
