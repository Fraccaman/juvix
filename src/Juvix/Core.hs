module Juvix.Core (
  module Juvix.Core.Dynamic,
  module Juvix.Core.Utility,
  module Juvix.Core.Types
) where

import           Juvix.Core.Dynamic
import           Juvix.Core.Types
import           Juvix.Core.Utility
