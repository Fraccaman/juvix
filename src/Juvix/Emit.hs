module Juvix.Emit where

import qualified Data.Text    as T
import           Foundation
import qualified Prelude      as P

import           Juvix.Script

emitUT ∷ ExprUT → T.Text
emitUT expr =
  case expr of

    UnitUT       → "UNIT"
    StringUT s   → T.concat ["\"", s, "\""]
    BoolUT b     → T.pack $ P.show b
    IntegerUT i  → T.pack $ P.show i
    TezUT t      → T.pack $ P.show t

    DropUT       → "DROP"
    DupUT        → "DUP"
    SwapUT       → "SWAP"
    ConstUT UnitUT        → "UNIT"
    ConstUT (BoolUT b)    → T.concat ["PUSH bool ", if b then "True" else "False"]
    ConstUT (IntegerUT i) → T.concat ["PUSH int ", T.pack (P.show i)]
    ConstUT (TezUT i)     → T.concat ["PUSH tez \"", T.pack (P.show i), "\""]
    ConstUT v             → emitUT v

    ConsPairUT   → "PAIR"
    CarUT        → "CAR"
    CdrUT        → "CDR"

    ConsSomeUT   → "SOME"
    ConsNoneUT   → "NONE"
    IfNoneUT a b → T.concat ["IF_NONE {", emitUT a, "} {", emitUT b, "}"]

    LeftUT       → "LEFT"
    RightUT      → "RIGHT"
    IfLeftUT a b → T.concat ["IF_LEFT {", emitUT a, "} {", emitUT b, "}"]

    ConsListUT   → "CONS"
    NilUT        → "NIL"
    IfConsUT a b → T.concat ["IF_CONS {", emitUT a, "} {", emitUT b, "}"]
    ListMapUT    → "MAP"
    ListReduceUT → "REDUCE"

    EmptySetUT   → "EMPTY_SET"
    SetMapUT     → "MAP"
    SetReduceUT  → "REDUCE"
    SetMemUT     → "MEM"
    SetUpdateUT  → "UPDATE"
    SetSizeUT    → "SIZE"

    EmptyMapUT   → "EMPTY_MAP"
    MapMapUT     → "MAP"
    MapReduceUT  → "REDUCE"
    MapMemUT     → "MEM"
    MapGetUT     → "GET"
    MapUpdateUT  → "UPDATE"
    MapSizeUT    → "SIZE"

    ConcatUT     → "CONCAT"

    AddSecondsToTimestampUT → "ADD"
    AddTimestampToSecondsUT → "ADD"

    AddTezUT     → "ADD"
    SubTezUT     → "SUB"
    MulTezNatUT  → "MUL"
    MulNatTezUT  → "MUL"
    EdivTezNatUT → "DIV"
    EdivTezUT    → "DIV"

    OrUT         → "OR"
    AndUT        → "AND"
    XorUT        → "XOR"
    NotUT        → "NOT"

    NegNatUT     → "NEG"
    NegIntUT     → "NEG"
    AbsNatUT     → "ABS"
    AbsIntUT     → "ABS"
    IntNatUT     → "CAST"
    AddIntIntUT  → "ADD"
    AddIntNatUT  → "ADD"
    AddNatIntUT  → "ADD"
    AddNatNatUT  → "ADD"
    SubIntUT     → "SUB"
    MulIntIntUT  → "MUL"
    MulIntNatUT  → "MUL"
    MulNatIntUT  → "MUL"
    MulNatNatUT  → "MUL"
    EdivIntIntUT → "DIV"
    EdivIntNatUT → "DIV"
    EdivNatIntUT → "DIV"
    EdivNatNatUT → "DIV"
    LslNatUT     → "LSL"
    LsrNatUT     → "LSR"
    OrNatUT      → "OR"
    AndNatUT     → "ADD"
    XorNatUT     → "XOR"
    NotNatUT     → "NOT"
    NotIntUT     → "NOT"

    SeqUT a b    → T.concat ["{", emitUT a, "; ", emitUT b, "}"]
    IfUT a b     → T.concat ["IF {", emitUT a, "} {", emitUT b, "}"]
    LoopUT a     → T.concat ["LOOP {", emitUT a, "}"]
    DipUT a      → T.concat ["DIP {", emitUT a, "}"]
    ExecUT       → "EXEC"
    LambdaUT a   → T.concat ["LAMBDA {", emitUT a, "}"]
    FailUT       → "FAIL"
    NopUT        → "NOP"

    CompareUT    → "COMPARE"
    EqUT         → "EQ"
    NeqUT        → "NEQ"
    LtUT         → "LT"
    GtUT         → "GT"
    LeUT         → "LE"
    GeUT         → "GE"

    CmpLtUT      → "CMPLT"
    CmpLeUT      → "CMPLE"
    CmpEqUT      → "CMPEQ"
    CmpGeUT      → "CMPGE"
    CmpGtUT      → "CMPGT"

    ManagerUT        → "MANAGER"
    TransferTokensUT → "TRANSFER_TOKENS"
    CreateAccountUT  → "CREATE_ACCOUNT"
    DefaultAccountUT → "DEFAULT_ACCOUNT"
    CreateContractUT → "CREATE_CONTRACT"
    NowUT            → "NOW"
    BalanceUT        → "BALANCE"
    CheckSignatureUT → "CHECK_SIGNATURE"
    HUT              → "H"
    StepsToQuotaUT   → "STEPS_TO_QUOTA"
    SourceUT         → "SOURCE"
    AmountUT         → "AMOUNT"

emitFinal ∷ ExprUT → T.Text
emitFinal expr =
  let code =
        case expr of
          NopUT → ""
          _     → emitUT expr in
  T.concat ["{", code, "}"]

emitType ∷ Type → T.Text
emitType IntT        = "int"
emitType KeyT        = "key"
emitType TezT        = "tez"
emitType BoolT       = "bool"
emitType StringT     = "string"
emitType UnitT       = "unit"
emitType (PairT a b) = T.concat ["(pair ", emitType a, " ", emitType b, ")"]
emitType (LamT _ _)  = ""
