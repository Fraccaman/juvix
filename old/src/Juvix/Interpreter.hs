{-# LANGUAGE CPP       #-}
{-# LANGUAGE MagicHash #-}

module Juvix.Interpreter where

import           Control.Concurrent
import           Control.Monad
import           Control.Monad.IO.Class
import qualified DynFlags
import           Foundation
import qualified GHC
import qualified GHC.LanguageExtensions as LangExt
import qualified GHC.Paths              as Paths
import qualified GHC.Prim               as GHC
import qualified GHCi
import qualified GHCi.RemoteTypes       as GHCi
import qualified HscMain                as GHC
import qualified Linker                 as GHC
import qualified Outputable             as GHC
import qualified Prelude                as P

import           Juvix.Core.Utility

data Command
  = Load P.String
  | Eval P.String

data Response
  = Loaded
  | Evaluated P.String

newtype InterpreterM = InterpreterM ()

ghcVersion ∷ Int
ghcVersion = __GLASGOW_HASKELL__

loadAndRun ∷ IO (Command → IO Response)
loadAndRun = do
  cmdChan ← newChan
  rspChan ← newChan
  void $ forkIO $ runInterpreter cmdChan rspChan
  return $ \cmd → do
    writeChan cmdChan cmd
    readChan rspChan

{-
runInterpreter ∷ Chan Command → Chan Response → IO a
runInterpreter cmdChan respChan = do
  oldFlags ← GHC.runGhc (Just Paths.libdir) DynFlags.getDynFlags
  let newFlags      = (DynFlags.updOptLevel 0 oldFlags) {
                          GHC.packageFlags = [DynFlags.ExposePackage "juvix" (DynFlags.PackageArg "juvix") (DynFlags.ModRenaming True [])] <> DynFlags.packageFlags oldFlags,
                          GHC.ghcMode = GHC.CompManager,
                          GHC.ghcLink = GHC.LinkInMemory,
                          GHC.hscTarget = DynFlags.HscInterpreted
                        }
      setFlags      = []
      unsetFlags    = []
      newFlags'     = foldr (flip DynFlags.wopt_set) (foldr (flip DynFlags.wopt_unset) newFlags unsetFlags) setFlags
      setGFlags     = [GHC.Opt_AutoLinkPackages]
      unsetGFlags   = []
      newFlags''    = foldr (flip DynFlags.gopt_set) (foldr (flip DynFlags.gopt_unset) newFlags' unsetGFlags) setGFlags
      setOptions    = [LangExt.UnicodeSyntax, LangExt.RebindableSyntax]
      unsetOptions  = [LangExt.ImplicitPrelude]
      finalFlags    = foldr (flip DynFlags.xopt_set) (foldr (flip DynFlags.xopt_unset) newFlags'' unsetOptions) setOptions
  hscEnv ← GHC.newHscEnv finalFlags
  forever $ do
    cmd ← liftIO (readChan cmdChan)
    resp ← case cmd of
      Eval expr → do
        fval ← GHCi.EvalThis |<< (GHCi.mkFinalizedHValue hscEnv =<< GHCi.mkRemoteRef (GHCi.HValue (GHC.unsafeCoerce# expr)))
        GHCi.EvalComplete _ (GHCi.EvalSuccess [result]) ← GHCi.evalStmt hscEnv False fval
        value ← GHCi.wormhole finalFlags result
        return (Evaluated (P.show value))
    liftIO (writeChan respChan resp)
-}

runInterpreter ∷ Chan Command → Chan Response → IO a
runInterpreter cmdChan respChan = GHC.runGhc (Just Paths.libdir) $ do
  oldFlags ← DynFlags.getDynFlags
  --mod ← GHC.findModule (GHC.mkModuleName "Juvix") Nothing
  let newFlags      = (DynFlags.updOptLevel 0 oldFlags) {
                          GHC.packageFlags = [DynFlags.ExposePackage "juvix" (DynFlags.PackageArg "juvix") (DynFlags.ModRenaming True [])] <> DynFlags.packageFlags oldFlags,
                          GHC.ghcMode = GHC.CompManager,
                          GHC.ghcLink = GHC.LinkInMemory,
                          GHC.hscTarget = DynFlags.HscInterpreted
                        }
      setFlags      = []
      unsetFlags    = []
      newFlags'     = foldr (flip DynFlags.wopt_set) (foldr (flip DynFlags.wopt_unset) newFlags unsetFlags) setFlags
      setGFlags     = [GHC.Opt_AutoLinkPackages]
      unsetGFlags   = []
      newFlags''    = foldr (flip DynFlags.gopt_set) (foldr (flip DynFlags.gopt_unset) newFlags' unsetGFlags) setGFlags
      setOptions    = [LangExt.UnicodeSyntax, LangExt.RebindableSyntax]
      unsetOptions  = [LangExt.ImplicitPrelude]
      finalFlags    = foldr (flip DynFlags.xopt_set) (foldr (flip DynFlags.xopt_unset) newFlags'' unsetOptions) setOptions
  toLink ← GHC.setSessionDynFlags finalFlags
  hscEnv ← GHC.getSession
  GHC.setTargets []
  _ ← GHC.load GHC.LoadAllTargets
  liftIO $ GHC.linkPackages hscEnv toLink

  --Just inf ← GHC.getModuleInfo mod
  let mod = GHC.mainModIs newFlags
  --liftIO $ P.print $ GHC.showSDoc newFlags $ GHC.ppr mod

  {-
  GHC.addTarget (GHC.Target (GHC.TargetModule (GHC.mkModuleName "Juvix")) True Nothing)
  GHC.depanal [] True
  GHC.load GHC.LoadAllTargets
  -}

  {-
  summary ← GHC.getModSummary (GHC.mkModuleName "Juvix")
  parsed ← GHC.parseModule summary
  typechecked ← GHC.typecheckModule parsed
  GHC.loadModule typechecked
  liftIO $ P.print $ GHC.showSDoc newFlags $ GHC.ppr mod
  -}

  --GHC.setContext [GHC.IIModule (GHC.mkModuleName "Juvix")]
  --GHC.setContext [GHC.IIModule (GHC.mkModuleName "Foundation")]
  --GHC.setContext [GHC.IIModule (GHC.mkModuleName "Prelude")]

  forever $ do
    cmd ← liftIO (readChan cmdChan)
    resp ← case cmd of
      Eval expr → do
        compiled ← GHC.dynCompileExpr expr
        return (Evaluated (P.show compiled))
    liftIO (writeChan respChan resp)
