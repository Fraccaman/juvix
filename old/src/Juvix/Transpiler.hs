module Juvix.Transpiler where

import           Control.Monad.Except
import           Control.Monad.RWS.Strict
import qualified Data.Map                         as M
import qualified Data.Text                        as T
import           Data.Typeable
import           Foundation
import qualified Prelude                          as P

import qualified Juvix.Backends.Michelson         as M hiding (Map)
import           Juvix.Core
import           Juvix.Core.CompilerTypes
import qualified Juvix.Core.GHC                   as GHC
import           Juvix.Transpiler.CoreToExpr
import           Juvix.Transpiler.CoreToType
import           Juvix.Transpiler.ExprToMichelson
import           Juvix.Transpiler.PrettyPrint
import           Juvix.Transpiler.SimplifyExpr

moduleToMichelson ∷ GHC.CoreProgram → GHC.CoreModule → ([CompileLog M.Type], Either CompileError (M.SomeExpr, M.Type, M.Type, M.Type))
moduleToMichelson l m =
  let binds ∷ GHC.CoreProgram
      binds   = l `mappend` GHC.cm_binds m
      env ∷ M.Map T.Text GHC.CoreExpr
      env     = M.fromList (P.concatMap (\case (GHC.NonRec v e) → [(prettyPrintValue v, e)]; GHC.Rec l → fmap (first prettyPrintValue) l) binds)
      computation ∷ CompilerM (M.SomeExpr, M.Type, M.Type, M.Type) M.Type
      computation = do
        core ← case filter (\case (GHC.NonRec v _) | nameToTextSimple (GHC.varName v) == "main" → True; _ → False) binds of
                  GHC.NonRec _ e:_ → tell [FrontendToCore e (GHC.exprType e)] >> return e
                  _                → throwError MainFunctionNotFound
        resolved <- resolveVars core
        expr ← coreToExpr resolved
        exprType ← coreToType core
        case exprType of
          M.LamT start@(M.PairT paramTy startStorageTy) end@(M.PairT retTy endStorageTy) | startStorageTy == endStorageTy → do
            case (M.liftType paramTy, M.liftType startStorageTy, M.liftType retTy, M.liftType endStorageTy) of
              (DynamicType (Proxy ∷ Proxy paramTyLifted), DynamicType (Proxy ∷ Proxy startStorageTyLifted), DynamicType (Proxy ∷ Proxy retTyLifted), DynamicType (Proxy ∷ Proxy endStorageTyLifted)) → do
                simplified ← simplifyExpr expr
                michelson ← exprToMichelson simplified
                (M.SomeExpr (expr ∷ M.Expr (M.Stack a) (M.Stack b)), _) ← M.liftUntyped michelson (M.typeToStack start) (DynamicType (Proxy :: Proxy startStorageTyLifted))
                case (eqT ∷ Maybe (a :~: (M.Pair paramTyLifted startStorageTyLifted, ())), eqT ∷ Maybe (b :~: (M.Pair retTyLifted endStorageTyLifted, ()))) of
                  (Just Refl, Just Refl) → do
                    optimized ← M.optimize expr
                    return (M.SomeExpr optimized, paramTy, retTy, startStorageTy)
                  _ → do
                    throwError (NotYetImplemented (T.concat ["Cannot unify start/end stack types: start ", prettyPrintValue start, " / end ", prettyPrintValue end, " with compilation output type ", T.pack (P.show (typeOf (undefined ∷ b))), "; this is a bug in Juvix and should be reported"]))
          _ → throwError (NotYetImplemented (T.concat ["Invalid type for main function - must be ∷ (param, storage) → (ret, storage) but was instead ", prettyPrintValue exprType]))
      (res, _, logs) = runRWS (runExceptT computation) (Env env) [FuncResult]
  in (logs, res)
