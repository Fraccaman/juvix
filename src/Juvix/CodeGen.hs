module Juvix.CodeGen where

import           Control.Monad.Except
import           Control.Monad.Writer
import           Protolude

import           Idris.AbsSyntax
import           Idris.Core.TT            hiding (Name)
import           Idris.ElabDecls
import           Idris.Main
import           Idris.Options
import           IRTS.CodegenCommon
import           IRTS.Compiler
import           IRTS.Lang

import qualified Juvix.Backends.Michelson as M
import           Juvix.Lang
import           Juvix.Utility

data Opts = Opts {inputs :: [FilePath],
                  output :: FilePath }

showUsage ∷ IO ()
showUsage = do putText "A code generator which is intended to be called by the compiler, not by a user."
               putText "Usage: idris-codegen-juvix <ibc-files> [-o <output-file>]"
               exitWith ExitSuccess

getOpts ∷ IO Opts
getOpts = do xs <- getArgs
             return $ process (Opts [] "main.jvx") xs
  where
    process opts ("-o":o:xs) = process (opts { output = o }) xs
    process opts (x:xs)      = process (opts { inputs = x:inputs opts }) xs
    process opts []          = opts

codeGenSdecls ∷ CodeGenerator
codeGenSdecls ci = do
  putText $ "Output file : " <> show (outputFile ci)
  let decls = liftDecls ci
  putText $ "Number of decls: " <> show (length decls)
  let main = findMain decls
  putText $ "Main decl print: " <> show main
  let decl = snd main
  putText $ "Main decl prettified: " <> prettyPrintValue decl
  let LFun _ _ args body = decl
      expr = LLam args body
  putText $ "Main expr prettified: " <> prettyPrintValue expr
  case runWriter $ runExceptT (M.compileToMichelsonSourceFile expr ∷ (ExceptT M.CompilationError (Writer [M.CompilationLog]) Text)) of
    (res, logs) -> do
      mapM_ (putText . prettyPrintValue) logs
      case res of
        Right output -> do
          writeFile (outputFile ci) output
          putText "Compilation success!"
        Left err -> do
          putText $ "Error during transpilation: " <> prettyPrintValue err
          exitFailure

findMain ∷ [(Name, LDecl)] → (Name, LDecl)
findMain decls = let Just f = head $ filter (\(name, _) -> name == NS (UN "main") ["Main"]) decls in f

sdeclMain ∷ Opts → Idris ()
sdeclMain opts = do elabPrims
                    _ <- loadInputs (inputs opts) Nothing
                    mainProg <- elabMain
                    ir <- compile (Via IBCFormat "juvix") (output opts) (Just mainProg)
                    runIO $ codeGenSdecls ir

main ∷ IO ()
main = do opts <- getOpts
          if (null (inputs opts))
             then showUsage
             else runMain (sdeclMain opts)

{-
data CodegenInfo = CodegenInfo {
    outputFile    :: String
  , outputType    :: OutputType
  , targetTriple  :: String
  , targetCPU     :: String
  , includes      :: [FilePath]
  , importDirs    :: [FilePath]
  , compileObjs   :: [String]
  , compileLibs   :: [String]
  , compilerFlags :: [String]
  , debugLevel    :: DbgLevel
  , simpleDecls   :: [(Name, SDecl)] -- most low level
  , defunDecls    :: [(Name, DDecl)]
  , liftDecls     :: [(Name, LDecl)]
  , interfaces    :: Bool
  , exportDecls   :: [ExportIFace]
  , ttDecls       :: [(Name, TTDecl)]
  }
-}
