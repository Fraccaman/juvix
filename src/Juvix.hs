module Juvix (
  compileToTz,
  pprint
) where

import           Control.Monad
import qualified Data.Text         as T
import qualified Data.Text.IO      as T
import           Foundation
import qualified GHC
import qualified Prelude           as P

import           Juvix.Emit
import           Juvix.GHC
import           Juvix.PrettyPrint
import           Juvix.Script
import           Juvix.Transpiler
import           Juvix.Types
import           Juvix.Utility

compileToTz ∷ P.FilePath → Bool → IO (Either CompileError T.Text)
compileToTz fn log = do
  lib ← compileModule "Juvix/Lib.hs"
  aux ← P.concat |<< mapM compilePackage ["Data.Ord"]
  mod ← compileModule fn
  let mainType          = typeOf mod "main"
      (logs, compiled)  = moduleToMichelson (GHC.cm_binds lib P.++ aux) mod
  let code              = compiled
  when log $ P.mapM_ (T.putStrLn . pprint) logs
  return $ do
    code ← code
    (LamT (PairT paramTy storageTy) (PairT retTy _)) ← mainType
    return $ T.unlines [
      T.concat ["parameter ", emitType paramTy, ";"],
      T.concat ["return ", emitType retTy, ";"],
      T.concat ["storage ", emitType storageTy, ";"],
      T.concat ["code ", emitFinal code, ";"]
      ]
