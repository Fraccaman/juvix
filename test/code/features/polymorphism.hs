module Polymorphism where

import           Juvix.Lib

fst ∷ (a, b) → a
fst (x, _) = x

snd ∷ (a, b) → b
snd (_, y) = y

main ∷ ((String, String), (Int, String)) → (String, (Int, String))
main (param, storage) = (fst param, (fst storage, snd storage))
