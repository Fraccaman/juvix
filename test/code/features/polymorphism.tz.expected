parameter (pair string string);
return string;
storage (pair int string);
code {{{{{DUP; {{{DUP; CDR}; SWAP}; CAR}}; {{{{{{{{{{{{SWAP; DUP}; DIP {SWAP}}; {{{DUP; CDR}; SWAP}; CAR}}; {{{{{{SWAP; DUP}; DIP {SWAP}}; SWAP}; DROP}; SWAP}; DROP}}; {{{{{DIP {SWAP}; SWAP}; DUP}; DIP {{SWAP; DIP {SWAP}}}}; {{{DUP; CDR}; SWAP}; CAR}}; {SWAP; DROP}}}; PAIR}; {{{{SWAP; DUP}; DIP {SWAP}}; {{{DUP; CDR}; SWAP}; CAR}}; {SWAP; DROP}}}; PAIR}; SWAP}; DROP}; SWAP}; DROP}}; SWAP}; DROP}};
